xarray/core/dataarray.py

```diff
@@ -840,6 +840,99 @@ def swap_dims(self, dims_dict):
+            the result array). If a list (or tuple) of integers is passed,
+            multiple axes are inserted. In this case, dim arguments should be
+            same length list or None. If None is passed, all the axes will be
+            inserted to the start of the result array.
+
+        Returns
+        -------
+        expanded : same type as caller
+            This object, but with an additional dimension(s).
+
+        Raises
+        -------
+        ValueError:
+               If the length of axis and dim are different.
+               If the axis is a list containing identical integers
+               If axis is invalid (larger than the original dimension+1)
+        """
+        def create_dim_name(src_dims):
+            """ A function to automatically generate dimension name. """
+            for i in range(len(src_dims)+1):
+                dim = 'dim_'+str(i)
+                if dim not in src_dims:
+                    return dim
+
+        # Make sure user does not do `expand_dims(0)`
+        if isinstance(dim, int):
+            raise ValueError('dim should be str or sequence of strs or dict')
+
+        # Make axis and dims a pair of lists
+        if isinstance(dim, basestring):
+            dim = [dim]
+        if axis is not None and not isinstance(axis, (list, tuple)):
+            axis = [axis]
+
+        # infer None arguments.
+        if axis is None:
+            if dim is None:
+                axis = [0]
+                dim = [create_dim_name(self.dims) for a in axis]
+            else:
+                axis = list(range(len(dim)))
+        if dim is None:
+            dim = []
+            for a in axis:
+                dim.append(create_dim_name(list(self.dims) + dim))
+
+        # Error checking
+        # Make sure the length of axis and dims are identical
+        if len(dim) != len(axis):
+            raise ValueError('lengths of dim and axis should be identical.')
+        # make sure any of dim does not already exist
+        for d in dim:
+            if d in self.dims:
+                raise ValueError('{dim} already exists.'.format(dim=d))
+
+        # make sure there is no duplicates in axis or dims
+        if len(axis) != len(set(axis)):
+            raise ValueError('axis should not contain same values.')
+        if len(dim) != len(set(dim)):
+            raise ValueError('dims should not contain same values.')
+
+        # sort dim and axis, so that axis is in order.
+        axis, dim = zip(*sorted(zip(axis, dim)))
+        # dims of the result array. This will be passed to Variable.expand_dims
+        all_dims = list(self.dims)
+        [all_dims.insert(a, d) for a, d in zip(axis, dim)]
+
+        # If dims includes a label of a scalar variables,
+        # it will be promoted to a 1D coordinate consisting of a single value.
+        coords = OrderedDict(self.coords)
+        for d in dim:
+            if d in coords.keys():
+                coords[d] = np.atleast_1d(coords[d])
```

Comment:
Instead of calling `np.atleast_1d`, use something like `Variable.expand_dims` to ensure this plays nicely with with non-numpy arrays.